aws_lambda_powertools/utilities/idempotency/idempotency.py

```diff
@@ -0,0 +1,109 @@
+    event: Dict[str, Any],
+    context: LambdaContext,
+    persistence_store: BasePersistenceLayer,
+) -> Any:
+    """
+    Middleware to handle idempotency
+
+    Parameters
+    ----------
+    handler: Callable
+        Lambda's handler
+    event: Dict
+        Lambda's Event
+    context: Dict
+        Lambda's Context
+    persistence_store: BasePersistenceLayer
+        Instance of BasePersistenceLayer to store data
+
+    Examples
+    --------
+    **Processes Lambda's event in an idempotent manner**
+        >>> from aws_lambda_powertools.utilities.idempotency import idempotent, DynamoDBPersistenceLayer
+        >>>
+        >>> persistence_store = DynamoDBPersistenceLayer(event_key="body", table_name="idempotency_store")
+        >>>
+        >>> @idempotent(persistence_store=persistence_store)
+        >>> def handler(event, context):
+        >>>     return {"StatusCode": 200}
+    """
+
+    try:
+        # We call save_inprogress first as an optimization for the most common case where no idempotent record already
+        # exists. If it succeeds, there's no need to call get_record.
+        persistence_store.save_inprogress(event=event)
+    except ItemAlreadyExistsError:
+        try:
+            event_record = persistence_store.get_record(event)
+        except ItemNotFoundError:
+            logger.debug(
+                "An existing idempotency record was deleted before we could retrieve it. Proceeding with lambda "
+                "handler"
+            )
+            return _call_lambda(handler=handler, persistence_store=persistence_store, event=event, context=context)
+
+        if event_record.status == STATUS_CONSTANTS["EXPIRED"]:
+            logger.debug(
+                f"Record is expired for idempotency key: {event_record.idempotency_key}. Proceeding with lambda "
+                f"handler"
+            )
+            return _call_lambda(handler=handler, persistence_store=persistence_store, event=event, context=context)
+
+        if event_record.status == STATUS_CONSTANTS["INPROGRESS"]:
+            raise AlreadyInProgressError(
+                f"Execution already in progress with idempotency key: "
+                f"{persistence_store.event_key}={event_record.idempotency_key}"
+            )
+
+        if event_record.status == STATUS_CONSTANTS["COMPLETED"]:
+            return event_record.response_json_as_dict()
+
+    return _call_lambda(handler=handler, persistence_store=persistence_store, event=event, context=context)
+
+
+def _call_lambda(
+    handler: Callable, persistence_store: BasePersistenceLayer, event: Dict[str, Any], context: LambdaContext
+) -> Any:
+    """
+
+    Parameters
+    ----------
+    handler: Callable
+        Lambda handler
+    persistence_store: BasePersistenceLayer
+        Instance of persistence layer
+    event
+        Lambda event
+    context
+        Lambda context
+    """
+    try:
+        handler_response = handler(event, context)
```

Comment:
1. Let's say that lambda timeouts before the handler returns, you may end up with with partial information in the persistence store but with an executed (or partially executed) operation. Probably this is an use case that's need to be clearly documented.
1. If the persistence store has a problem (i.e. throttling) you may end up with a situation where the operation was finished, the status is never recorded, so once the expiration time passes the operation may be called again, as there was no record this has already been properly executed.

Reply:
The handler cannot start executing until a lock is obtained ("in progress" record in the persistence layer). If we can't obtain that lock, lambda_handler will not be executed. If the lock was obtained but the operation did not complete / Lambda times out, the "in progress" record will persist, preventing any further executions of the function with the same idempotency key. When the record expires - regardless of what status it was in the event can be retried: this will be the documented behavior.

I'm not sure there's much we can do (aside from clearly documenting the behavior) to handle partially executed operations where the lambda times out during the handler. If you have any improvement ideas on that they're very welcome!